libdrgn/kallsyms.c

```diff
@@ -0,0 +1,625 @@
+// Copyright (c) 2024 Oracle and/or its affiliates
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <ctype.h>
+#include <stddef.h>
+
+#include "kallsyms.h"
+#include "program.h"
+#include "drgn.h"
+#include "string_builder.h"
+#include "symbol.h"
+
+/**
+ * This struct contains the tables necessary to reconstruct kallsyms names.
+ *
+ * vmlinux (core kernel) kallsyms names are compressed using table compression.
+ * There is some description of it in the kernel's "scripts/kallsyms.c", but
+ * this is a brief overview that should make the code below comprehensible.
+ *
+ * Table compression uses the remaining 128 characters not defined by ASCII and
+ * maps them to common substrings (e.g. the prefix "write_"). Each name is
+ * represented as a sequence of bytes which refers to strings in this table.
+ * The two arrays below comprise this table:
+ *
+ *   - token_table: this is one long string with all of the tokens concatenated
+ *     together, e.g. "a\0b\0c\0...z\0write_\0read_\0..."
+ *   - token_index: this is a 256-entry long array containing the index into
+ *     token_table where you'll find that token's string.
+ *
+ * To decode a string, for each byte you simply index into token_index, then use
+ * that to index into token_table, and copy that string into your buffer.
+ *
+ * The actual kallsyms symbol names are concatenated into a buffer called
+ * "names". The first byte in a name is the length (in tokens, not decoded
+ * bytes) of the symbol name. The remaining "length" bytes are decoded via the
+ * table as described above. The first decoded byte is a character representing
+ * what type of symbol this is (e.g. text, data structure, etc).
+ */
+struct kallsyms_reader {
+	uint32_t num_syms;
+	uint8_t *names;
+	char *token_table;
+	uint16_t *token_index;
+	bool long_names;
+};
+
+/*
+ * Kallsyms doesn't include symbol length.  We determine symbol length by the
+ * start of the subsequent symbol.  Unfortunately, there can be large gaps in
+ * the symbol table, for instance on x86_64 the Linux kernel has percpu symbols
+ * near the beginning of the address space, and a large gap before normal kernel
+ * symbols.  The result of this is that we can create symbols with incredibly
+ * large sizes, and then drgn's symbolization will print addresses using that
+ * symbol and a very large offset, which is absolutely meaningless.
+ *
+ * To avoid this, we set a cap on the length of a symbol. Unfortunately, this is
+ * a heuristic. It's entirely possible to have very large data symbols. This
+ * value is chosen somewhat arbitrarily, but seems to produce decent results.
+ */
+#define MAX_SYMBOL_LENGTH 0x10000
```

Comment:
FWIW, here are the top 10 symbol sizes in a vmtest kernel, a few of which are over this threshold:

```
$ eu-readelf -s build/vmtest/x86_64/kernel-6.11.0-rc2-vmtest30.1default/build/vmlinux | awk '{ print $3 "\t" $8 }' | sort -rn | head
360448  _printk_rb_static_infos
262144  early_dynamic_pgts
131072  __log_buf
98304   _printk_rb_static_descs
65536   stack_pools
65536   __brk_early_pgt_alloc
65536   __brk_dmi_alloc
65536   __apicid_to_node
49536   i8042_dmi_quirk_table
32768   ftrace_stacks
```

I have no idea how useful those symbols are.

Reply:
The entire purpose of `MAX_SYMBOL_LENGTH` is to ensure we don't allow `__per_cpu_end` to have a size which extends across the entire userspace memory region (and possibly similar issues in other architectures). So I don't mind increasing this.

That said, I did notice that on my kernel (6.9 based) there are some kallsyms symbols that are now coming out of order:
```
ffffffffc028d4e0 t __pfx_fuse_invalidate_atime  [fuse]
ffffffffc02900f0 t __pfx_fuse_do_setattr        [fuse]
ffffffffc028f8a0 t fuse_reverse_inval_entry     [fuse]
```
This was unintentionally catching the overflow from the subtraction here. I've now explicitly caught that overflow. This only happens for module symbols -- vmlinux symbols are in order. I don't really have any solution for this other than not using `/proc/kallsyms` for module kallsyms: getting it from `struct module` is better.